cpp/include/raft/neighbors/detail/nn_descent_batch.cuh

```diff
@@ -0,0 +1,684 @@
+                          raft::host_vector_view<IdxT, IdxT> inverted_indices,
+                          raft::host_vector_view<IdxT, IdxT> cluster_size,
+                          raft::host_vector_view<IdxT, IdxT> offset)
+{
+  // build sparse inverted indices and get number of data points for each cluster
+  size_t num_rows = global_nearest_cluster.extent(0);
+  size_t k        = global_nearest_cluster.extent(1);
+
+  auto local_offset = raft::make_host_vector<IdxT>(n_clusters);
+
+  max_cluster_size = 0;
+  min_cluster_size = std::numeric_limits<size_t>::max();
+
+  for (size_t i = 0; i < n_clusters; i++) {
+    cluster_size.data_handle()[i] = 0;
+    local_offset.data_handle()[i] = 0;
+  }
+
+  for (size_t i = 0; i < num_rows; i++) {
+    for (size_t j = 0; j < k; j++) {
+      IdxT cluster_id = global_nearest_cluster.data_handle()[i * k + j];
+      cluster_size.data_handle()[cluster_id]++;
+    }
+  }
+
+  offset.data_handle()[0] = 0;
+  for (size_t i = 1; i < n_clusters; i++) {
+    offset.data_handle()[i] = offset.data_handle()[i - 1] + cluster_size.data_handle()[i - 1];
+  }
+  for (size_t i = 0; i < num_rows; i++) {
+    for (size_t j = 0; j < k; j++) {
+      IdxT cluster_id = global_nearest_cluster.data_handle()[i * k + j];
+      inverted_indices
+        .data_handle()[offset.data_handle()[cluster_id] + local_offset.data_handle()[cluster_id]] =
+        i;
+      local_offset.data_handle()[cluster_id]++;
+    }
+  }
+
+  for (size_t i = 0; i < n_clusters; i++) {
+    if (max_cluster_size < cluster_size.data_handle()[i]) {
+      max_cluster_size = cluster_size.data_handle()[i];
+    }
+    if (min_cluster_size > cluster_size.data_handle()[i]) {
+      min_cluster_size = cluster_size.data_handle()[i];
+    }
+  }
+}
+
+template <typename KeyType, typename ValueType>
+struct KeyValuePair {
+  KeyType key;
+  ValueType value;
+};
+
+template <typename KeyType, typename ValueType>
+struct CustomKeyComparator {
+  __device__ bool operator()(const KeyValuePair<KeyType, ValueType>& a,
+                             const KeyValuePair<KeyType, ValueType>& b) const
+  {
+    if (a.key == b.key) { return a.value < b.value; }
+    return a.key < b.key;
+  }
+};
+
+template <typename IdxT, int BLOCK_SIZE, int ITEMS_PER_THREAD>
+RAFT_KERNEL merge_subgraphs(IdxT* cluster_data_indices,
+                            size_t graph_degree,
+                            size_t num_cluster_in_batch,
+                            float* global_distances,
+                            float* batch_distances,
+                            IdxT* global_indices,
+                            IdxT* batch_indices)
+{
+  size_t batch_row = blockIdx.x;
+  typedef cub::BlockMergeSort<KeyValuePair<float, IdxT>, BLOCK_SIZE, ITEMS_PER_THREAD>
+    BlockMergeSortType;
+  __shared__ typename cub::BlockMergeSort<KeyValuePair<float, IdxT>, BLOCK_SIZE, ITEMS_PER_THREAD>::
+    TempStorage tmpSmem;
+
+  extern __shared__ float sharedMem[];
```

Comment:
Byte-based array as you reinterpret the shared memory as different types below. You might have to switch the shared memory calculation accordingly
```suggestion
  extern __shared__ char sharedMem[];
```